File: shell-sort/shell-sort.ts
function shellSort(arr: number[]) {
  const arrCpy = arr.slice();
  const length = arrCpy.length + 1;
  const f = Math.floor;

  let pointer = f(length / 2);
  while (pointer > 0) {
    for (let i = pointer; i < length - 1; i++) {
      let tmp = arrCpy[i];
      let y = i;
      for (; y >= pointer && arrCpy[y - pointer] > tmp; y = y - pointer)
        arrCpy[y] = arrCpy[y - pointer];
      arrCpy[y] = tmp;
    }
    pointer = f(pointer / 2);
  }

  return arrCpy;
}

const input = [9, 5, 2, -125, 5, 6, 1, -4, -6, 7, 214, 516, 103, 0, 13];
console.log("Input:", input, "\n");
console.log("Output:", shellSort(input));

File: shell-sort/shell-sort.c
#include <stdio.h>

void shellSort(int arr[], int n) {
  int gap = n / 2;

  while (gap > 0) {
    for (int i = gap; i < n; i += 1) {
      int tmp = arr[i], x = i;
      for (; x >= gap && arr[x - gap] > tmp; x -= gap) arr[x] = arr[x - gap];
      arr[x] = tmp;
    }

    gap /= 2;
  }
}

int main() {
  int arr[] = {9, 8, 7, 6, 5, 4, 3, 2, 1, 0};
  int n = sizeof(arr) / sizeof(arr[0]);

  printf("Input\n");
  for (int i : arr) printf("%i,", i);

  shellSort(arr, n);

  printf("\nOutput\n");
  for (int i : arr) printf("%i,", i);

  return 0;
}
